lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs

```diff
@@ -260,6 +267,32 @@ simpleSpec cp = do
             runExceptT (putCheckpoint db testPk cp) `shouldReturn` Right ()
             readCheckpoint db testPk `shouldReturn` Just cp
 
+-- | Can rollback to checkpoints created after
+prop_rollbackApplyBlocks
+    :: (Eq s)
+    => DBLayer IO s DummyTarget k
+    -> Wallet s DummyTarget
+    -> SlotAndBlocks
+    -> Property
+prop_rollbackApplyBlocks db state (SlotAndBlocks (slotTo, blocks)) =
+    monadicIO $ do
+    cps <- setup
+    let cp:_ = filter (\s -> let (BlockHeader slot _ _) = currentTip s
+                             in slot == slotTo) cps
+    prop cp
+  where
+    setup = liftIO $ do
+        unsafeRunExceptT $ createWallet db testPk state testMetadata
+        let addCheckpoint cps@(cp:_) block = do
+                let (_,cp') = applyBlock block cp
+                runExceptT (putCheckpoint db testPk cp') `shouldReturn` Right ()
+                pure $ cp':cps
+            addCheckpoint [] _ = pure []
```

Comment:
Hmmm. Perhaps using `NonEmpty` would avoid the weird `addCheckpoint [] _` branch here :+1: